_posts/2021-10-06-jvm-issues-at-comcast.md

```diff
@@ -0,0 +1,219 @@
+
+First, we looked at the queries involved in these nightly reports. We
+noticed that all these queries involved 2 particular tables. In this post,
+we will call them table A and table B.
+
+Our initial suspicion was that there could be an issue with the table data in
+HDFS. Thus, we tried to reproduce the performance problem by using queries that
+performed simple scans against these tables. 
+
+We tried a simple table scan with no filters, range filter on a partitioned
+column, etc.,  ran these queries multiple times and execution times were
+consistent. This ruled out a potential problem with HDFS.
+
+Next, we took a closer look  at the portion of the slow running queries
+involving the A table and came up with the simplest possible query that could
+demonstrate the problem. We discovered that the following query performed as
+expected:
+
+```
+select
+ count(a.c1)
+from
+ hive.schema1.A a, hive.schema2.B da
+where
+ a.day_id = da.date_id
+ and a.day_id BETWEEN '2021-03-22' AND '2021-04-21'
+```
+
+But adding a predicate caused the performance problem to appear:
+
+```
+select
+ count(a.c1)
+from
+ hive.schema1.A a, hive.schema2.date_dim da
+where
+ a.day_id = da.date_id
+ and a.day_id BETWEEN '2021-03-22' AND '2021-04-21'
+ and a.c2 = '4 (Success)'
+```
+
+We narrowed the problem down to the `Scan/Filter/Project` operator using the
+output of `EXPLAIN ANALYZE` from Trino. For the query that performed as
+expected, this stage had the following CPU stats:
+
+```
+CPU: 2.39h, Scheduled: 4.47h, Input: 17434967615 rows (357.47GB)
+```
+
+For the simple version of the query performing poorly, the same stage has the
+following CPU stats:
+
+```
+CPU: 3.73d, Scheduled: 48.01d, Input: 17052985227 rows (413.98GB)
+```
+
+This shows that for roughly the equivalent amount of data, Trino used
+significantly more CPU (3.73 days to 2.39 hours!!). Our next step was to
+determine why that was and exactly what was using additional CPU.
+
+We generated a few jstack and JFR profiles of the Starburst Java process from
+one of the worker nodes while the scan stage was running. After analyzing these
+profiles, we found no obvious problem; Trino performed as expected.
+
+Next, we looked at the list of tasks in the web UI to see what the distribution
+of CPU times for each stage was:
+
+![](/assets/blog/jvm-issues-at-comcast/web_ui_before.png)
+
+Notice in the above screenshot, some workers have tasks that only use up a few
+minutes of CPU time and others have tasks that use up to 2 hours of CPU time!
+Different query runs would show this would happen to different workers so it
+was not a problem with any one individual worker.
+
+We discussed this with Starburst engineer, [Piotr Findeisen](https://github.com/findepi),
+and came to the conclusion that this could potentially be an issue with JVM
+code deoptimization. After re-compiling a method a certain number of times,
+the JVM will refuse to do so any more and will run the method in interpreted
```

Comment:
change to present tense

the JVM refuses .. and runs in the much slower interpreted mode.